modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/inline/InlineTreeFilterClosure.java

```diff
@@ -56,12 +60,20 @@ public InlineTreeFilterClosure(IndexingQueryCacheFilter cacheFilter,
         if (!val)
             return false;
 
-        if (rowFilter != null)
+        if (rowFilter != null) {
             val = rowFilter.apply(tree, io, pageAddr, idx);
 
+            lastRow = rowFilter.lastRow();
+        }
+
         return val;
     }
 
+    /** {@inheritDoc} */
+    @Override public @Nullable IndexRow lastRow() {
+        return lastRow;
```

Comment:
Do we really need a field here? 
Can we use `rowFilter == null ? null : rowFilter.lastRow()` here?

Reply:
We actually need `lastRow`, in case a row is filtered by `cacheFilter` it should return `null`.